app/models/user.rb

```diff
@@ -0,0 +1,40 @@
+class User < ApplicationRecord
+  # self note: the dependent: :delete_all allows for a work to be deleted even though a user and a work are connected using foreign keys in the votes table
+  has_many :votes, dependent: :delete_all
+
+  validates :username, presence: true, uniqueness: true
+
+  def vote(input_work_id)
+    vote = Vote.where(user_id: self.id, work_id: input_work_id)
+    # vote = Vote.find_by(user_id: self.id, work_id: input_work_id)
+    return vote
+  end
+
+  def array_of_votes
+    votes = Vote.where(user_id: self.id)
+    return votes
+  end
+
+  def works_upvoted_ids
+    upvotes = array_of_votes
+    works_ids = []
+
+    upvotes.each do |upvote|
+      works_ids << upvote.work_id
+    end
+
+    return works_ids
+  end
+
+  def works_voted_on
+    works_ids = works_upvoted_ids
+    works = []
+
+    works_ids.each do |work_id|
+      work = Work.find_by(id: work_id)
+      works << work
+    end
+
+    return works
+  end
```

Comment:
The only time the method `works_upvoted_ids` is used is within this method (and the tests)... would it also work if we combined the two?

If we combine the two methods above, we'll see the following code:

```ruby
  def works_voted_on

    upvotes = array_of_votes
    works_ids = []

    upvotes.each do |upvote|
      works_ids << upvote.work_id
    end

    works = []

    works_ids.each do |work_id|
      work = Work.find_by(id: work_id)
      works << work
    end

    return works
  end
```

When we see that code, we can see some repetition... Instead of iterating through `upvotes` (the result of `array_of_votes`) AND also iterating through `work_ids` (the result of iterating through `upvotes`), could we combine it to a new refactoring:

```ruby
  def works_voted_on

    upvotes = array_of_votes
    works = []

    upvotes.each do |upvote|
      works << upvote.work
    end

    return works
  end
```

If this works, it relies on us understanding that ActiveRecord and the relationship we set up for `Vote`s (such that a `Vote` belongs to a `Work`), and that each `Vote` can call `vote.work` to get a `Work`. It relies on us feeling a little more comfortable with making shortcuts, but it sure does make the code shorter!